File: Codinator/Project/Universal Classes/File Mover/FileMoverViewController.m
//
//  ExportViewController.m
//  VWAS-HTML
//
//  Created by Vladimir on 22.12.14.
//  Copyright (c) 2014 Vladimir Danila. All rights reserved.
//

#import "FileMoverViewController.h"
#import "AppDelegate.h"

@interface FileMoverViewController (){
    
    NSString *filename;
    NSString *oldFilePath;
}

@property (nonatomic, strong) AppDelegate *appDelegate;


@end

@implementation FileMoverViewController
@synthesize path_, rootPath;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self update];
    
    
    oldFilePath = path_;
    filename  = path_.lastPathComponent;
    
    path_ = rootPath;
    
    
    items_ = [[[NSFileManager defaultManager] contentsOfDirectoryAtPath:path_ error:nil] mutableCopy];
    
    self.appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



- (NSInteger)numberOfSectionsInTableView:(UITableView *)aTableView {
    // Return the number of sections.
    return 1;
}


- (NSInteger)tableView:(UITableView *)aTableView numberOfRowsInSection:(NSInteger)section {
    return [items_ count];
}

- (UITableViewCell *)tableView:(UITableView *)tableViewFunc cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *CellIdentifier = @"CellIdentifier";
    
    UITableViewCell *cell = [tableViewFunc dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
        cell.backgroundColor = [UIColor blackColor];
        cell.tintColor = [UIColor purpleColor];
    }
    
    // Configure the cell.
    

    
    NSString *fileName = items_[indexPath.row];
    cell.textLabel.text = fileName;
    cell.textLabel.textColor = [UIColor lightGrayColor];
    
    
    NSString *path = [path_ stringByAppendingPathComponent:fileName];
    cell.imageView.image = [self.appDelegate.thumbnailManager thumbnailForFileAtPath:path];
    
    
    
    return cell;
    
    
}




- (void)tableView:(UITableView *)aTableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSString *path = [path_ stringByAppendingPathComponent:items_[indexPath.row]];

    BOOL isDir;
    
    [[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDir];
    
    if (isDir) {
        
        backButton.enabled = YES;
        
        path_ = path;
        NSError *error;
        items_ = [[[NSFileManager defaultManager] contentsOfDirectoryAtPath:path_ error:&error] mutableCopy];
        [tableView reloadData];
    }


    
    
}


- (IBAction)importButton:(id)sender {

    NSString *newFilePath = [path_ stringByAppendingPathComponent:filename];
    
    [[NSFileManager defaultManager] moveItemAtPath:oldFilePath toPath:newFilePath error:nil];
    
    
    [self dismissViewControllerAnimated:true completion:^{
        [[NSNotificationCenter defaultCenter] postNotificationName:@"history" object:self userInfo:nil];
    }];
}




- (IBAction)backButton:(id)sender {
    NSError *error;
    NSString *newPath = [path_ stringByDeletingLastPathComponent];
    

    
    if ([rootPath isEqualToString:newPath]) {
        backButton.enabled = NO;
    }
    else{
    path_ = newPath;
    items_ = [[[NSFileManager defaultManager] contentsOfDirectoryAtPath:path_ error:&error] mutableCopy];
    [tableView reloadData];
    }
}

-(void)update{

    
    if ([path_ isEqualToString:rootPath]) {
        backButton.enabled = NO;
    }

    
    [self performSelector:@selector(update) withObject:nil afterDelay:0.08];
    
}

@end

File: Codinator/Project/Universal Classes/File Mover/FileMoverViewController.swift
//
//  FileMoverViewController.swift
//  Codinator
//
//  Created by Vladimir Danila on 25/04/16.
//  Copyright © 2016 Vladimir Danila. All rights reserved.
//

import UIKit

class FileMoverViewController: UIViewController, UITableViewDelegate, UITableViewDataSource{

    @IBOutlet weak var tableView: UITableView!
    
    var fileUrl: NSURL?
    
    var items: [NSURL]?
    
    let fileManager = NSFileManager.defaultManager()
    
    
    var delegate: NewFilesDelegate?
    
    @IBOutlet weak var backButton: UIBarButtonItem!
    var backButtonEnabled: Bool {
        get {
            print(inspectorUrl?.path)
            return !(inspectorUrl!.path!.hasSuffix(".cnProj/Assets") || inspectorUrl!.path!.hasSuffix(".cnProj/Assets/"))
        }
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()

        do {
            inspectorUrl = fileUrl!.URLByDeletingLastPathComponent!
            items = try fileManager.contentsOfDirectoryAtURL(inspectorUrl!, includingPropertiesForKeys: [], options: .SkipsHiddenFiles)
            tableView.reloadData()
            
        } catch let error as NSError {
            Notifications.sharedInstance.alertWithMessage(error.localizedDescription, title: "Error", viewController: self)
            self.dismissViewControllerAnimated(true, completion: nil)
        }
    }

    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        backButton.enabled = backButtonEnabled
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: - Buttons
    
    @IBAction func cancelDidPush() {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    @IBAction func moveFile() {
        
        let destinationUrl = inspectorUrl?.URLByAppendingPathComponent(fileUrl!.lastPathComponent!)
        if fileUrl?.absoluteString != destinationUrl?.absoluteString {
            
            do {
                try fileManager.moveItemAtURL(fileUrl!, toURL: destinationUrl!)
                self.dismissViewControllerAnimated(true, completion: {
                    self.delegate?.reloadData()
                })
            } catch let error as NSError {
                Notifications.sharedInstance.alertWithMessage(error.localizedDescription, title: "Error", viewController: self)
            }
        
        }
        else {
            Notifications.sharedInstance.alertWithMessage(nil, title: "Locations are the same", viewController: self)
        }
    
    }
    
    @IBAction func backDidPush(sender: UIBarButtonItem) {
        inspectorUrl = inspectorUrl?.URLByDeletingLastPathComponent
        
        do {
            items = try fileManager.contentsOfDirectoryAtURL(inspectorUrl!, includingPropertiesForKeys: [], options: .SkipsHiddenFiles)
            tableView.reloadData()
            
        } catch let error as NSError {
            Notifications.sharedInstance.alertWithMessage(error.localizedDescription, title: "Error", viewController: self)
            self.dismissViewControllerAnimated(true, completion: nil)
        }
        
        backButton.enabled = backButtonEnabled
    }
    

    // MARK: - Table View
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items!.count
    }
    
    
    var inspectorUrl: NSURL?
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath)

        
        
        let bgColorView = UIView()
        bgColorView.backgroundColor = UIColor.blackColor()
        cell.selectedBackgroundView = bgColorView
        cell.backgroundColor = tableView.backgroundColor
        
        
        if let text = items![indexPath.row].lastPathComponent {
            cell.textLabel?.text = text
            cell.textLabel?.textColor = UIColor.whiteColor()
            
            if let path = inspectorUrl?.URLByAppendingPathComponent(text) {
                let manager = Thumbnail()
                cell.imageView?.image = manager.thumbnailForFileAtPath(path.path)
            }
        }
        
        
        
        return cell
    }
    
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        let selectedUrl = inspectorUrl?.URLByAppendingPathComponent(items![indexPath.row].lastPathComponent!, isDirectory: true)
        
        
        var isDirectory : ObjCBool = ObjCBool(false)
        
        if (NSFileManager.defaultManager().fileExistsAtPath(selectedUrl!.path!, isDirectory: &isDirectory) && Bool(isDirectory) == true) {
            
            do {
                inspectorUrl = selectedUrl
                items = try fileManager.contentsOfDirectoryAtURL(inspectorUrl!, includingPropertiesForKeys: [], options: .SkipsHiddenFiles)
                tableView.reloadData()
                
            } catch let error as NSError {
                Notifications.sharedInstance.alertWithMessage(error.localizedDescription, title: "Error", viewController: self)
                self.dismissViewControllerAnimated(true, completion: nil)
            }
            
            backButton.enabled = backButtonEnabled
            
        }
        
    }
    
}
